panel/panel.py

```diff
@@ -136,64 +118,51 @@ def run_update_panel_cmd(panel, text=None):
     panel.run_command(cmd, {'text': text, 'clear_sel': clear_sel})
 
 
-def format_row(lineno, error_type, dic):
-    lineno = int(lineno) + 1
-    start = dic['start'] + 1
-    msg = dic['msg'].rstrip()
-    tmpl = " {LINENO:>5}:{START:<4} {ERR_TYPE:7} {linter:>12}: {code:12} {MSG}"
-    return tmpl.format(
-        LINENO=lineno, START=start, ERR_TYPE=error_type, MSG=msg, **dic)
+def format_row(item):
+    line = int(item["line"]) + 1
+    start = item['start'] + 1
+    tmpl = " {LINE:>5}:{START:<4} {error_type:7} {linter:>12}: {code:12} {msg}"
+    return tmpl.format(LINE=line, START=start, **item)
 
 
-def filter_ok(check_val, key, panel_filter):
+def passes_listing(buf_d, panel_filter):
     """Check value against white- and blacklisting and return bool."""
     # check whitelisting
     if not panel_filter:
         return True
 
-    if key in panel_filter and check_val not in panel_filter[key]:
-        return False
+    x = {"types": "error_type", "codes": "code", "linter": "linter"}
+    for key, check_val in x.items():
+        # check whitelisting
+        if key in panel_filter:
+            if buf_d[check_val] not in panel_filter[key]:
+                return False
 
-    # check blacklisting
-    exclude_key = "exclude_" + key
-    if exclude_key in panel_filter and check_val in panel_filter[exclude_key]:
-        return False
+        # check blacklisting
+        exclude_key = "exclude_" + key
+        if exclude_key in panel_filter:
+            if buf_d[check_val] in panel_filter[exclude_key]:
+                return False
 
+    # if all checks passed return True
     return True
 
 
-def get_view_lines(view_dict, panel_filter):
-    view_lines = []
-    for lineno, line_dict in sorted(view_dict["line_dicts"].items()):
-        for error_type in WARN_ERR:
-            if not filter_ok(error_type, "types", panel_filter):
-                continue
-
-            err_dict = line_dict.get(error_type)
-            if not err_dict:
-                continue
-            items = sorted(err_dict, key=lambda k: k['start'])
-
-            items = [
-                item for item in items
-                if filter_ok(item['linter'], "linter", panel_filter) and
-                filter_ok(item['code'], "codes", panel_filter)
-            ]
-            for item in items:
-                view_lines.append((lineno, error_type, item))
-
-    return view_lines
+def get_buf_lines(buf_dicts, panel_filter):
+    buf_dicts = [d for d in buf_dicts if passes_listing(d, panel_filter)]
+    return sorted(buf_dicts, key=lambda x: (x["line"], x["start"], x["end"]))
```

Comment:
Line redundant? Are `raw_errors` already sorted that way?

Reply:
No. Unsorted, arbitrary order.